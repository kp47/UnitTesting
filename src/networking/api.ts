import axios, {AxiosRequestConfig, AxiosResponse} from 'axios';

axios.defaults.headers.post['Content-Type'] = 'application/json';
axios.defaults.headers.post.Accept = 'application/json';

const config: AxiosRequestConfig = {
  baseURL: 'api/login',
  timeout: 50000,
};

const instance = axios.create(config);

// Interceptors

const requestInterceptorId: number = instance.interceptors.request.use(
  (config: AxiosRequestConfig) => config,
  (error: any) => Promise.reject(error),
);

instance.interceptors.request.use(
  async (config: AxiosRequestConfig) => Promise.resolve(config),
  (error: any) => Promise.reject(error),
);

instance.interceptors.request.eject(requestInterceptorId);

instance.interceptors.request.use(
  (config: AxiosRequestConfig) => Promise.resolve(config),
  (error: any) => Promise.reject(error),
);

instance.interceptors.response.use(
  response => response,
  async (error: any) => {
    const {status} = error.response;
    // console.log('Error Status: ', status);
    return Promise.reject(new Error('Incorrect Credentials'));
  },
);

instance.interceptors.request.use((config: AxiosRequestConfig) => config);
instance.interceptors.request.use((config: AxiosRequestConfig) =>
  Promise.resolve(config),
);

const responseInterceptorId: number = axios.interceptors.response.use(
  (response: AxiosResponse) => response,
  (error: any) => Promise.reject(error),
);

axios.interceptors.response.eject(responseInterceptorId);

axios.interceptors.response.use(
  (response: AxiosResponse) => Promise.resolve(response),
  (error: any) => {
    Promise.reject(error);
  },
);

axios.interceptors.response.use((response: AxiosResponse) => response);
axios.interceptors.response.use((response: AxiosResponse) =>
  Promise.resolve(response),
);

export interface AxiosProps {
  type: any;
  url: any;
  params?: any;
  headers?: any;
}

/** function that returns an axios call */
export const requestApi = ({
  type,
  url,
  params,
  headers,
}: AxiosProps): Promise<AxiosResponse> => {
  return instance.request({
    method: type,
    url: url,
    data: params,
    headers: headers,
  });
};

export default instance;
