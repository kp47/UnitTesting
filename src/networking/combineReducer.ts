import {combineReducers, Reducer} from '@reduxjs/toolkit';
import loginReducer from '../redux/login/reducer';

const rootReducer: Reducer<any, any> = combineReducers({
  loginReducer,
});

export default rootReducer;
