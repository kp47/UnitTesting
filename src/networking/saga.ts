import {all, fork} from 'redux-saga/effects';

import watchOnLoginScreenRequest from '../redux/login/saga';

export default function* rootSaga() {
  yield all([fork(watchOnLoginScreenRequest)]);
}
