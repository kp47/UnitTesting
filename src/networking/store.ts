import {applyMiddleware, createStore} from 'redux';
import createSagaMiddleware from 'redux-saga';
import persistedReducer from './combineReducer';
import sagas from './saga';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(persistedReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(sagas);

export {store, sagaMiddleware};
