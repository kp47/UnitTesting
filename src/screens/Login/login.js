import {useRoute} from '@react-navigation/native';
import React, {useEffect, useState} from 'react';
import {Image, Text, TextInput, TouchableOpacity, View} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {getLoginScreenRequest} from '../../redux/login/action';
import {
  // isLoading,
  getAllLoginScreenError,
  getAllLoginScreenSuccess,
} from '../../redux/login/selector';
import styles from './styles';

const Login = () => {
  const route = useRoute();
  const {name} = route.params;
  const dispatch = useDispatch();
  const [error, setError] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const token: any = useSelector(getAllLoginScreenSuccess);
  const LoginScreenListError: any = useSelector(getAllLoginScreenError);
  // const loader: any = useSelector(isLoading);

  useEffect(() => {
    LoginScreenListError && setError('Username & Password do not match.');
    if (token) {
      console.log('token', token?.token);
      setError('');
    }
  }, [LoginScreenListError, token]);

  const validate = () => {
    dispatch(
      getLoginScreenRequest({
        email: email,
        password: password,
      }),
    );
  };

  return (
    <View>
      <View style={styles.logo}>
        <Image
          source={require('../../images/react_native.png')}
          style={styles.images}
        />
        <Text testID={'name'} style={[styles.login]}>
          Welcome {name}
        </Text>
        <Text testID={'errorMessage'} style={[styles.login, styles.errorColor]}>
          {error}
        </Text>
      </View>
      <View style={styles.form}>
        <TextInput
          testID={'userName'}
          style={styles.inputFormItems}
          placeholder="Email "
          onChangeText={text => setEmail(text)}
          value={email}
        />
        <TextInput
          testID={'password'}
          style={styles.inputFormItems}
          placeholder="Password "
          secureTextEntry={true}
          onChangeText={text => setPassword(text)}
          value={password}
        />
        <TouchableOpacity activeOpacity={0.9}>
          <Text style={[styles.login, styles.fpStyle]}>
            {'Forgot Password?'}
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          testID={'login'}
          onPress={() => validate()}
          style={styles.register}>
          <Text style={styles.text}>Login</Text>
        </TouchableOpacity>
      </View>
      <View style={styles.mainLinks}>
        <TouchableOpacity activeOpacity={0.9}>
          <Image
            source={require('../../images/facebook.png')}
            style={styles.links}
          />
        </TouchableOpacity>
        <TouchableOpacity activeOpacity={0.9}>
          <Image
            source={require('../../images/google.png')}
            style={styles.links}
          />
        </TouchableOpacity>
        <TouchableOpacity activeOpacity={0.9}>
          <Image
            source={require('../../images/twitter.png')}
            style={styles.links}
          />
        </TouchableOpacity>
      </View>
      <TouchableOpacity activeOpacity={0.9}>
        <Text style={styles.login}>
          {"Don't have an account?"}
          <Text style={styles.boldText}>{' Register'}</Text>
        </Text>
      </TouchableOpacity>
    </View>
  );
};

export default Login;
