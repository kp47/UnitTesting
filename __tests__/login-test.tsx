/**
 * @format
 */

import {render, fireEvent, act, waitFor} from '@testing-library/react-native';
import React from 'react';
import {Provider} from 'react-redux';
import {store} from '../src/networking/store';
import 'react-native';
import Login from '../src/screens/Login/login';

describe('Login Screen page test', () => {
  test('Check to see if renders properly', async () => {
    const component = (
      <Provider store={store}>
        <Login />
      </Provider>
    );

    const wrapper = render(component);
    expect(wrapper).toBeTruthy();
    expect(wrapper.toJSON()).toMatchSnapshot();

    const textToEnter = 'InvalidUsername';
    // const textToEnter = 'eve.holt@reqres.in';
    const input = await wrapper.findByTestId('userName');
    fireEvent.changeText(input, textToEnter);

    const passwordToEnter = 'InvalidPassword';
    // const passwordToEnter = 'cityslicka';
    const passwordInput = await wrapper.findByTestId('password');
    fireEvent.changeText(passwordInput, passwordToEnter);

    await act(async () => fireEvent.press(wrapper.getByTestId('login')));
    wrapper.rerender(component);

    // wait for appearance inside an assertion
    await waitFor(() => {
      expect(
        wrapper.getByText('Username & Password do not match.'),
      ).toBeTruthy();
    });
  });
});
