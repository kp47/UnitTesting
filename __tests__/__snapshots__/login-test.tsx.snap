// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen page test Check to see if renders properly 1`] = `
<View>
  <View
    style={
      Object {
        "height": "45%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/react_native.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 200,
          "width": 200,
        }
      }
    />
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "#3590e6",
            "fontWeight": "600",
            "marginTop": 20,
          },
        ]
      }
      testID="name"
    >
      Welcome 
      Kunal
    </Text>
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "#3590e6",
            "fontWeight": "600",
            "marginTop": 20,
          },
          Object {
            "color": "red",
          },
        ]
      }
      testID="errorMessage"
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "marginHorizontal": 10,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Email "
      style={
        Object {
          "borderColor": "#39c6ed",
          "borderWidth": 0.5,
          "fontSize": 18,
          "marginHorizontal": 30,
          "marginTop": 10,
          "padding": 10,
        }
      }
      testID="userName"
      value=""
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Password "
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "#39c6ed",
          "borderWidth": 0.5,
          "fontSize": 18,
          "marginHorizontal": 30,
          "marginTop": 10,
          "padding": 10,
        }
      }
      testID="password"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "alignSelf": "center",
              "color": "#3590e6",
              "fontWeight": "600",
              "marginTop": 20,
            },
            Object {
              "alignSelf": "flex-end",
              "marginRight": 30,
            },
          ]
        }
      >
        Forgot Password?
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#55f2dd",
          "borderRadius": 5,
          "height": 50,
          "marginHorizontal": 40,
          "marginTop": 30,
          "opacity": 1,
          "paddingBottom": 10,
          "paddingTop": 10,
          "width": "85%",
        }
      }
      testID="login"
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/images/facebook.png",
          }
        }
        style={
          Object {
            "alignItems": "center",
            "marginTop": 30,
            "paddingBottom": 5,
            "paddingTop": 10,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/images/google.png",
          }
        }
        style={
          Object {
            "alignItems": "center",
            "marginTop": 30,
            "paddingBottom": 5,
            "paddingTop": 10,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/images/twitter.png",
          }
        }
        style={
          Object {
            "alignItems": "center",
            "marginTop": 30,
            "paddingBottom": 5,
            "paddingTop": 10,
          }
        }
      />
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#3590e6",
          "fontWeight": "600",
          "marginTop": 20,
        }
      }
    >
      Don't have an account?
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
         Register
      </Text>
    </Text>
  </View>
</View>
`;
